crates/noirc_evaluator/src/brillig/brillig_ir/artifact.rs

```diff
@@ -0,0 +1,162 @@
+/// nature cannot be fully resolved while building the bytecode either.
+/// We add relative jumps into the `Relative` variant of this enum.
+#[derive(Debug, Clone)]
+pub(crate) enum UnresolvedJumpLocation {
+    Label(String),
+    Relative(i32),
+}
+
+impl BrilligArtifact {
+    /// Link two Brillig artifacts together and resolve all unresolved jump instructions.
+    pub(crate) fn link(&mut self, obj: &BrilligArtifact) -> Vec<BrilligOpcode> {
+        self.append_artifact(obj);
+        self.resolve_jumps();
+        self.byte_code.clone()
+    }
+
+    /// Link with an external brillig artifact.
+    ///
+    /// This method will offset the positions in the Brillig artifact to
+    /// account for the fact that it is being appended to the end of this
+    /// Brillig artifact (self).
+    fn append_artifact(&mut self, obj: &BrilligArtifact) {
+        let offset = self.index_of_next_opcode();
+        for (jump_label, jump_location) in &obj.unresolved_jumps {
+            self.unresolved_jumps.push((jump_label + offset, jump_location.clone()));
+        }
+
+        for (label_id, position_in_bytecode) in &obj.labels {
+            self.labels.insert(label_id.clone(), position_in_bytecode + offset);
+        }
+
+        self.byte_code.extend_from_slice(&obj.byte_code);
+    }
+
+    /// Adds a brillig instruction to the brillig byte code
+    pub(crate) fn push_opcode(&mut self, opcode: BrilligOpcode) {
+        self.byte_code.push(opcode);
+    }
+
+    /// Adds a unresolved jump to be fixed at the end of bytecode processing.
+    pub(crate) fn add_unresolved_jump(
+        &mut self,
+        jmp_instruction: BrilligOpcode,
+        destination: UnresolvedJumpLocation,
+    ) {
+        assert!(
+            Self::is_jmp_instruction(&jmp_instruction),
+            "expected a jump instruction, but found {jmp_instruction:?}"
+        );
+
+        self.unresolved_jumps.push((self.index_of_next_opcode(), destination));
+        self.push_opcode(jmp_instruction);
+    }
+
+    /// Returns true if the opcode is a jump instruction
+    fn is_jmp_instruction(instruction: &BrilligOpcode) -> bool {
+        matches!(
+            instruction,
+            BrilligOpcode::JumpIfNot { .. }
+                | BrilligOpcode::JumpIf { .. }
+                | BrilligOpcode::Jump { .. }
+        )
+    }
+
+    /// Adds a label in the bytecode to specify where this block's
+    /// opcodes will start.
+    pub(crate) fn add_label_at_position(&mut self, label: String, position: OpcodeLocation) {
+        let old_value = self.labels.insert(label.clone(), position);
+        assert!(
+            old_value.is_none(),
+            "overwriting label {label}. old_value = {old_value:?}, new_value = {position}"
+        );
+    }
+
+    /// Returns the index of the next opcode.
+    ///
+    /// This is useful for labelling regions of code
+    /// before you have generated the opcodes for the region.
+    pub(crate) fn index_of_next_opcode(&self) -> usize {
```

Comment:
```suggestion
    pub(crate) fn index_of_next_opcode(&self) -> OpcodeLocation {
```
Nit, might be a bit more clear